### DISCUSSION TITLE
Changes to grub configs

### DISCUSSION THREAD

**Discussion by Itxaka** [rancher/elemental-toolkit#1647] [Ideas]
Coming from https://github.com/rancher/elemental-cli/pull/373

We have an issue when searching for files from grub in some server FW which makes the search take 30 minutes.

The idea here would be to on install with elemental-cli create those files if they dont exists so the search is not stuck.

After some discussions it also come to light some better approach to our grub.cfg configurations:

 - merge grubenv and grub_oem_env into just one file
 - merge grubmenu and grubcustom into just one file
 - stop searching everywhere for those files and instead set them to be in the STATE partition always
 - That means no more uncertainty when loading them (first found first loaded no more, that is not stable)
 - elemental-cli will check for those files on install/upgrade/reset and create empty files if not found so grub is not stalled
 - examples will be added on how to modify/customize those files in after_install/after_upgrade

**Comment by Itxaka** [comment]
cc @bk201 @davidcassany @mudler 

This will only affect you if:
 - You are using grubenv for extra grub configs and not grub_oem_env (consolidation of both files into grub_oem_env)
 - You are using grubmenu for extra menus and not grubcustom (consolidation of both files into grubcustom)
 - Store grubenv/grub_oem_env/grubmenu/grubcustom in a different partition than STATE

**Comment by Itxaka** [comment]
Doesnt seem that great tbh, there is no way to override anything without a complex yip file and random mounting/unmounting the state partition which yuck...

Another approach would be:

 - use one file for env one for menus
 - Store them in the state partition by default
 - Allow the same files to be on the OEM partition which would override the default ones
 

This allows us to have some static files in there AND allow overriding by end users easily. Not a great approach but at least we cna streamline this.

**Reply by davidcassany** [reply]
yes, I was afraid of the amount of changes that could imply, specially in our tests.

Another probably not straight forward approach could be providing a new elemental-cli command for these files. So we could add some additional logic to prevent insane approaches or having to instruct users to do crazy remounts. For variables we could also van changes for certain ones and full freedom only for new ones. A command that could be used in both after-install hooks and in init stages. At least of the grubenv I believe this could be useful and make sense, not entirely sure for the custom config file though.

Imagine something like

```
$ elemental setgrubvars myvar=myvalue next_entry=recovery 
```

How does it sounds?

**Reply by Itxaka** [reply]
blergh, sounds bad to be honest. Having a command to remount the state partition just for a single line change...I dont want to think what could go wrong, plus the whole remounting RW and cleaning up...so much stuff could go wrong. IMHO the less we touch the RO partitions the best, thats why I hate it by going with the override via OEM sounds the best approach. Not that I like it thougth.... :crying_cat_face:

**Reply by davidcassany** [reply]
I see your point, probably as @kkaempf suggested we should step back and define our implementation goals (such as prevent as much as possible searching for files) and our feature set (include additional stuff in grub.cfg). Then probably having this clarified we can discuss about backward compatibility contstraints. I'll try to do such an exercise in this discussion later in the morning.

**Reply by Itxaka** [reply]
yes, good point. Will close the current open PR until we can discuss this and clarify everything.

**Comment by Itxaka** [comment]
We need:
 - No search

We want:
 - Less files
 - No meddling with mounting/umounting partitions -> danger
 - Allow users to override values for grub (Which values? During install/upgrade only as we control the flow? Or even in a running system?


I think having the default files on STATE, even if empty its good. Then we can let the users override by having the same files on the OEM partition, but instead of searching for them blindly, first load STATE then test if those files are also in OEM and load those.
This gets us, no searching, a set of defaults written to RO disk (so they always exist) and flexibility to override.

Although I don't really like that end users can easily break the system by writing to the OEM partition (RW by default) as that could make the whole thing flimsy. Probably need to investigate further but its a real possibility that you cannot boot anymore if you break the grub menus/configs somehow ¬_¬

At least you can probably insert a USB with a COS_OEM partition and have them load it from there :rofl: 

At a minimum I would do the load default from STATE, then check for  overrides on OEM instead of the actual "it will load from whenever it finds them and does a random search for those files, no matter your disk layout". Would even leave the rest as is, then afterwards merge the files. That would meet our needs for no searching and give the process a bit more stability while keeping the overridable feature and enhancing the user docs

**Comment by davidcassany** [comment]
I believe the main question to answer here is how to handle a grub configuration that allows grub2 customization at the same time we keep a simple and _immutable or almost completely static_ grub configuration. Tried to dump and organize here all the ideas I have around this topic.

### About use cases (feature set)

For that I would divide customization use cases into three categories:

1. **Build time customization**: this customization happens at OS build time, basically forking our current configuration or even re-implementing it completely. This is a **not relevant use case for us**, this is already fully possible and not supported from a product POV.
2. **Permanent customization**: this type of customization, in general, is expected to happen only once before/during/after the grub installation.
    * add extra menu
    * set menu display name
    * set filesystem labels
    * set boot arguments
    * set default menu entry
    * include extra grub.cfg logic for extra features (boot assessment)

    Those are the type of things that are expected to not change across reboots, not related to any runtime state. Note that during install/ugrade time we have already mounted in rw mode the relevant partitions.
3. **Runtime time customization**: this is what requires runtime state to be functional and meaningful. Examples could be:
    * set next reboot menu option, only once.
    * set and keep traces of previous boot failures

### About implementation goals

* Make grub configuration file (`grub.cfg`) as static as possible from build time going through install time up to run time. Ideally grub.cfg should not be templated (currently it is at build time to handle filesystem labels).
* Keep it simple (short and without too many files around)
* Avoid inefficient grub features such as search for files as much as possible (depending on the hardware setup this can be tricky)
* Make grub configuration upgradable as much as possible


### Eventual constraints

* Keeping secure boot is a must

In terms of compatibility:
* Backward compatibility: That is, given a current grub installation we can upgrade OS to a newer version and it will boot. So the new OS is backward compatible with previous grub configurations.
* Forward compatibility: That is, given a current grub installation we can upgrade to an older OS and it will boot. So the old OS is forward compatible with the new grub configuration.

## Ideas for implementation

* Any sort of runtime state stat should be tracked in OEM partition as grub2 environment variables, no extra config files should be used. Just pure and mere variables. (i.e. track boot failures or configure default menu options for only once)

* Permanent customization should be divided in two:
    1. In can be set at runtime (applied in OEM)
        * additional boot arguments
        * default menu entry
        * menu display name
    3. It can be set at install/upgrade time only (applied in STATE partition)
        * add logic from config files (e.g. boot assessment extra logic)
        * add extra menus
        * set layout related data such as labels

For that I would set a `grub.cfg` such as:

```bash
# Set defaults
set timeout=10
set display_name="Elemental"
set fallback="0 1 2"

# Load system env file (includes label variables set during install)
# no need to include device in path, default device is already the state partition
# also the $root variable could be exported from the EFI config
# see https://www.gnu.org/software/grub/manual/grub/grub.html#export
set system_env="/grub2/system_grubenv"
load_env -f "${system_env}"

# Load user env file
set user_env="/grub2/user_grubenv"
search --no-floppy --label --set=oem_blk "${oem_label}"
if [ "${oem_blk}" ] ; then
  load_env -f "(${oem_blk})${user_env}"
  # Reload state to prevent overloads from OEM
  load_env -f "${system_env}"
fi

# Save default
if [ "${next_entry}" ]; then
  set default="${next_entry}"
  set selected_entry="${next_entry}"
  set next_entry=
  save_env -f "(${env_blk})${user_env}" next_entry
else
  set default="${saved_entry}"
fi

insmod all_video
insmod loopback
insmod squash4

menuentry "${display_name}" --id cos {
  # state label is taken from system env
  # no need to set the active label, more over its not even needed to boot,
  # having the image path and state label is sufficient it this was redundant
  set img=/cOS/active.img
  loopback loop0 /$img
  source (loop0)/etc/elemental/bootargs.cfg
  linux (loop0)$kernel $kernelcmd ${extra_active_cmdline}
  initrd (loop0)$initramfs
}

menuentry "${display_name} (fallback)" --id fallback {
...
}

menuentry "${display_name} recovery" --id recovery {
...
}

# Load extra config, custom_blk defined in system_env
# I think this should be at the bottom to ensure the fallback variable is
# consistent with the number in case new menu blocks are added here
if [ "${custom_file}" ]; then
  source "${custom_file}"
fi
```

Then the `/etc/elemental/bootargs.cfg` file included into the image could be something like:

```bash
set kernel=/boot/vmlinuz
# This could also boot squashfs images [^_^]
# ${default_args} could be set at install time (system_env) and include
# some defaults like "console=tty1 console=ttyS0 fsck.mode=force fsck.repair=yes"
set kernelcmd="root=LABEL=$state_label cos-img/filename=$img panic=5 rd.cos.oemlabel=$oem_label ${default_args}"
set initramfs=/boot/initrd
```

We could consider making `/grub2/system_grubenv` configurable/extendable at install time via the yaml config to modify `${default_args}` for boot parameters (addresses rancher/elemental-cli#297). 

With this setup:

* `grub.cfg` is no longer templated at build time, we could finally drop `packages/cos-config`
* some simple customization is still possible without touching ro partitions (displayname, next_entry, default_entry, fallback, timeout) by just setting env vars into a file in OEM (and only in OEM)
* it requires some explicit intervention to break grub config (install time hooks or runtime manual rw remounts)
* wider customization is possible at install time setting the custom file (via after-install hook) and/or setting variables at a variables file (`/grub2/system_env`) in state partition.
* `grub.cfg` is slightly shortly than it was.
* at a glance I'd say this setup, in essence, can be backward compatible with nowadays setup
* at a glance I'd say it has potential to be forward compatible under some constraints (e.g. no squash)
* the is not a single search of device by file only OEM and Recovery would be searched by label
* `grub.cfg` is absolutely transparent to squashfs
* no additional modules, secure boot is kept


### Onwards compatibility contsraints

This setup defines three defacto contracts to keep compatibility across updates:

* `elemental install` must set certain grub variables at install time in `/grub2/system_env` (only labels are mandatory `${state_label}`, `${oem_label}` and `${recovery_label}`)
* `/etc/elemental/bootargs.cfg` must exist in image and define `${kernel}`, `${kernelcmd}` and `${initramfs}`
* `grub.cfg` must define `${img}` for active, fallback and recovery menu entries.

### Gray areas

I believe the only relevant gray area is how the boot assessment needs to be adapted to this setup:

* _Installation_ should happen at install/upgrade time via a hook
* Runtime state should all be kept as a grub env variables in OEM (it is unclear to me if this leaves some some corner cases uncovered)
* Also unclear to me if adding extra grub config logic before the menu entries is mandatory
